### ISSUE TITLE
lines suspected O(n²)

### ISSUE THREAD

**Issue by nh2** [snoyberg/conduit#341]
I suspect that https://github.com/snoyberg/conduit/blob/33feb2d32da4a0564d01b6c61481b7e5f8084597/conduit-extra/Data/Conduit/Text.hs#L87-L101 is O(n²) if I send it one char after the other.

Ideas:

* Switch to `Text.Lazy` for `T.append` to avoid quadratic append
* Don't search (T.`break`) from the beginning each time, only searc for `\n` in the newly `await`ed chunk

**Comment by nh2** [comment]
I propose:

```haskell
import qualified Data.Text.Lazy as TL

-- | Split input by the given car, emitting each line separately.
linesSeparatedBy :: Monad m => Char -> Conduit Text m Text
linesSeparatedBy sep =
    awaitText TL.empty
  where
    awaitText buf = do
      await >>= \mTextChunk -> case mTextChunk of
        Nothing -> unless (TL.null buf) (yield (TL.toStrict buf))
        Just chunk ->
          let (lineEnd, rest) = T.break (== sep) chunk
          -- `rest` is empty or starts with `;`
          in  case T.uncons rest of
                Nothing ->
                  -- Here `lineEnd == chunk` and `rest == ""`
                  awaitText (buf `TL.append` TL.fromStrict chunk)
                Just (_semicolon, rest') -> do
                  yield $ TL.toStrict (buf `TL.append` TL.fromStrict lineEnd)
                  yieldLines rest'

    yieldLines text =
      let (line, rest) = T.break (== sep) text
      in  case T.uncons rest of
            Just (_, rest') -> yield line >> yieldLines rest'
            _ -> awaitText (TL.fromStrict line)

lines :: Monad m => Conduit Text m Text
lines = linesSeparatedBy '\n'
```

This should be linear (though I wrote it between 02:00 and 03:00 so review appreciated). If it's considered good, I'll send a PR, but let's scrutinise it first.

The idea:

* Always search for a `\n` in the newly awaited `chunk` only. Never search in `buf`.
* When a `\n` appears in `buf`, a line is complete and we can yield it (`buf ++ thatBitBeforeTheNewline`).
* There may be more newlines in the `chunk`, so before continuing, yield other lines in `theBitsAfterTheNewline`.

Invariants:

1. Always use `TL.append` for O(1) appends.
2. Always search for a `\n` in the newly awaited `chunk :: Text` only. Never search in `buf`.
3. Any `toStrict` result will be immediately `yield`ed.

Complexity:

* By (2), the amount of data going through `T.break` is O(n). 
* By (3), the amount of copying is O(n).

**Comment by nh2** [comment]
> I suspect that 

I've actually tested now that it is quadratic (instead of just eye-balling it):

```haskell
> length $ runConduitPure $ CL.sourceList [ "c" | _ <- [1..100000::Int] ] .| CT.lines .| CL.consume
1
```

takes a minute or so in my ghci.

Using my proposed version:

```haskell
> length $ runConduitPure $ CL.sourceList [ "c" | _ <- [1..100000::Int] ] .| linesSeparatedBy '\n' .| CL.consume
1
```

finishes immediately.

**Comment by snoyberg** [comment]
Addressed by #338
